### PR TITLE
Prevent infinite redirect loops for subsequent CAS responses

### PR DESCRIPTION
**Pull request by kkatpcc**
Added capability and config option (CASAllowSubAuth) that allows subsequent CAS authentication responses to pass through, leaving the ticket parameter intact. Helps prevent infinite redirect loops when CAS protection is being used at multiple levels.

See https://github.com/apereo/mod_auth_cas/issues/144 for original inspiration and prototype code.

### CODE REVIEW

**Comment by kkatpcc** [comment]
I initially thought of names like "preserve" or "pass-through" but realized those could be just as ambiguous due to context. It would only preserve or pass-through on subsequent requests after initial authentication. Hence, SubAuth. Overall, I figured there would be discussion on a better name in this PR, so I left it at that. If anyone can think of a name that would make this concept less ambiguous overall, I am all for it!

I'll fix the logic so that it adds the headers. I will probably just move the common logic into a separate function and use it in both places.

**Comment by dhawes** [comment]
> I'll fix the logic so that it adds the headers. I will probably just  move the common logic into a separate function and use it in both  places.

That's exactly what I'd suggest.

The README should be updated as well.

**Comment by kkatpcc** [comment]
Code & README updated as requested.

Just need to conclude the directive renaming party.

File: README [hunk 1]
```diff
@@ -296,6 +296,13 @@ Description:	If enabled, the response from the CAS Server will be parsed for SAM
 		CASValidateURL appropriately; typical URLs are of the form
 		https://login.example.org/cas/samlValidate.
 
+Directive:	CASAllowSubAuth
+Default:	Off
+Description:	This directive determines whether subsequent CAS authentication
+		responses are allowed to pass through, leaving the ticket parameter intact.
```
Comment (dhawes) [inline README]: I don't like the phrase "CAS authentication responses". 

Is something like the following more accurate:

```This directive leaves CAS ticket parameters intact when a valid session cookie exists. This helps prevent infinite redirect loops when CAS protection is being used at multiple levels.```

File: src/mod_auth_cas.c [hunk 2]
```diff
@@ -2081,6 +2113,16 @@ int cas_authenticate(request_rec *r)
 	ticket = getCASTicket(r);
 	cookieString = getCASCookie(r, (ssl ? d->CASSecureCookie : d->CASCookie));
 
+	// prevent infinite redirect loops by allowing subsequent authentication responses to pass through, leaving the ticket parameter intact
+	if(c->CASAllowSubAuth && (ticket != NULL) && (cookieString != NULL) && ap_is_initial_req(r) && isValidCASCookie(r, c, cookieString, &remoteUser, &attrs)) {
+		cas_set_attributes(r, attrs);
+		r->user = remoteUser;
```
Comment (dhawes) [inline src/mod_auth_cas.c]: Should we check that remoteUser exists here?

Comment (kkatpcc) [reply src/mod_auth_cas.c]: Doesn't seem to have an affect either way during basic testing, but I added the check to match the other code.

Comment (dhawes) [reply src/mod_auth_cas.c]: I'm more concerned when a remoteUser doesn't exist.

I wonder if this should mirror https://github.com/portlandcc/mod_auth_cas/blob/allow_sub_auth/src/mod_auth_cas.c#L2236.

If remoteUser could be NULL, I don't think returning OK is the right thing to do. I haven't done any analysis, but it is needed.

Comment (dhawes) [reply src/mod_auth_cas.c]: If the remoteUser is NULL, something went wrong reading or writing the cookie, which seems unlikely. If we want to be super safe, it's probably not a bad idea to add it to the outer if.

Comment (kkatpcc) [reply src/mod_auth_cas.c]: The remoteUser NULL check was moved to the outer if.

**Comment by dhawes** [comment]
Ahh, the name game.

The reason I like something like CASPreserveTicket is that it's general enough to cover multiple use cases. Your specific use case is to use CAS at multiple levels. Another use case could be to pass a ticket parameter to something like an issue tracker or whatever application also uses the ticket parameter.

CASPreserveTicketForValidSession. Ugh. ;)

**Comment by serac** [comment]
I was going to make a name suggestion, but I honestly can't come up with anything better. I personally don't like "CASAllowSubAuth" because "SubAuth" means absolutely nothing to me at face value. I might be able to guess what "CASTicketPassthrough" did given the name, but I admit it's not terribly clear. If I had to pick from among the options thus far, I think I'd go for "CASPreserveTicket". I think documentation could fill any gaps about how it works.

**Comment by kkatpcc** [comment]
I'm fine with _CASPreserveTicket_ as the directive name. Nothing will really substitute for RTFM anyhow.

@dhawes: Let me know if you want to leave the directive naming decision open a little longer, or if we have enough consensus already for renaming it to _CASPreserveTicket_.

**Comment by dhawes** [comment]
Let's go with CASPreserveTicket.

**Comment by kkatpcc** [comment]
Done. The _CASAllowSubAuth_ directive is now called _CASPreserveTicket_.

**Pull request merged**